### PR TITLE
V3

### PR DESCRIPTION
**Pull request by prateek3255**
This update is a complete rewrite of the react-otp-input component internally and has a few breaking changes. The updates include:

- The earlier version was written with class components which is not recommended anymore and plain JSX leading to no typesafety. To fix this update is rewritten with React hooks and TypeScript offering complete typesafety and much cleaner code.
- Got rid of the `onInput` since `onChange` supersedes it and is recommended by React instead.
- Add a new renderInput prop that now requires the user to pass their own input and hence allows them to style it or add custom props to it according to their needs, this also helps get rid of many other props that were being passed to the component and then passed to the input, which is not needed anymore because they can be directly passed to the input. Apart from that this also has a few more breaking changes which can be found in the [migration guide](https://github.com/devfolioco/react-otp-input/tree/v3#migrating-from-v2).
- The library is now compiled via rollup instead of webpack, [since it is the more preferred bundler for libraries and is more easier to use.](https://medium.com/webpack/webpack-and-rollup-the-same-but-different-a41ad427058c)
- The example is also rewritten using the Vite react template, which offers lightning fast hot reloads.

### CODE REVIEW

File: src/index.tsx [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+    let nextActiveInput = activeInput;
+
+    // Get pastedData in an array of max size (num of inputs - current position)
+    const pastedData = event.clipboardData
+      .getData('text/plain')
+      .slice(0, numInputs - activeInput)
+      .split('');
+
+    // Prevent pasting if the clipboard data contains non-numeric values
+    if (inputType === 'number' && pastedData.some((value) => isNaN(Number(value)))) {
```
Comment (gitryder) [inline src/index.tsx:190]: @prateek3255 I guess we should also be checking for `tel` too here

**Review by gitryder — APPROVED** [review]
🚀

File: src/index.tsx (hunk 1 above)
Comment (prateek3255) [reply src/index.tsx:190]: Yes thanks for pointing out

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    "noImplicitReturns": true,
+    "noImplicitThis": true,
+    "noImplicitAny": true,
+    "strictNullChecks": true,
+    "suppressImplicitAnyIndexErrors": true,
+    "allowSyntheticDefaultImports": true
+  },
+  "include": ["src"],
+  "exclude": ["node_modules"]
+}
```
Comment (preetjdp) [inline tsconfig.json:22]: Missing new line at the end of the file

image

**Review by preetjdp — APPROVED** [review]
Minor Nit; Apart from that LGTM

**Comment by prateek3255** [comment]
https://github.com/all-contributors please add @prateek3255 for code, maintenance, ideas, docs, design

**Comment by prateek3255** [comment]
@all-contributors please add @prateek3255 for code, maintenance, ideas, docs, design

**Pull request merged**